src/names/scopes.md

```diff
@@ -1,3 +1,329 @@
+        // break 'a;
+    }
+}
+
+// The label is in scope for the expression of `while` loops.
+'a: while break 'a {}         // Loop does not run.
+'a: while let _ = break 'a {} // Loop does not run.
+
+// The label is not in scope in the defining `for` loop:
+'a: for outer in 0..5 {
+    // This will break the outer loop, skipping the inner loop and stopping
+    // the outer loop.
+    'a: for inner in { break 'a; 0..1 } {
+        println!("{}", inner); // This does not run.
+    }
+    println!("{}", outer); // This does not run, either.
+}
+
+```
+
+Loop labels may shadow labels of the same name in outer scopes.
+References to a label refer to the closest definition.
+
+```rust
+// Loop label shadowing example.
+'a: for outer in 0..5 {
+    'a: for inner in 0..5 {
+        // This terminates the inner loop, but the outer loop continues to run.
+        break 'a;
+    }
+}
+```
+
+> **Note**: `rustc` currently warns about defining the same loop label multiple times in the same function, or if the label has the same name as an in-scope lifetime.
+> This is intended as a future-compatibility warning about a possible extension to the language.
+> See [PR #24162](https://github.com/rust-lang/rust/pull/24162).
+
+## Prelude scopes
+
+[Preludes] bring entities into scope of every module.
+The entities are not members of the module, but are implicitly queried during [name resolution].
+The prelude names may be shadowed by declarations in a module.
+
+The preludes are layered such that one shadows another if they contain entities of the same name.
+The order that preludes may shadow other preludes is the following where earlier entries may shadow later ones:
+
+1. [Extern prelude]
+2. [Tool prelude]
+3. [`macro_use` prelude]
+4. [Standard library prelude]
+5. [Language prelude]
+
+## `macro_rules` scopes
+
+The scope of `macro_rules` macros is described in the [Macros By Example] chapter.
+The behavior depends on the use of the [`macro_use`] and [`macro_export`] attributes.
+
+## Derive macro helper attributes
+
+[Derive macro helper attributes] are in scope in the item where their corresponding [`derive` attribute] is specified.
+The scope extends from just after the `derive` attribute to the end of the item. <!-- Note: Not strictly true, see https://github.com/rust-lang/rust/issues/79202, but this is the intention. -->
+Helper attributes shadow other attributes of the same name in scope.
+
+## `Self` scope
+
+Although [`Self`] is a keyword with special meaning, it interacts with name resolution in a way similar to normal names.
+
+The implicit `Self` type in the definition of a [struct], [enum], [union], [trait], or [implementation] is treated similarly to a [generic parameter](#generic-parameter-scopes), and is in scope in the same way as a generic type parameter.
+For structs, enums, and unions, it is in scope starting after the generic parameters.
+For traits and implementations, it is in scope starting just before the generic parameters.
```

Comment:
I'm not sure that this difference is not a bug.
The original implementation of `Self` is for traits and impls, and it is certainly correct.

Reply:
I think I made a mistake here, I'm not sure what I was thinking at the time.  Perhaps I got confused by [this code](https://github.com/rust-lang/rust/blob/f9b2e3c87b76c03cc398be8b8f65d096c0583eb2/compiler/rustc_resolve/src/late.rs#L1082-L1094), but based on [this](https://github.com/rust-lang/rust/blob/f9b2e3c87b76c03cc398be8b8f65d096c0583eb2/compiler/rustc_resolve/src/late.rs#L1532-L1554) I'm almost certain that struct/enum/union have the same behavior as traits and impls.


```rust
struct S<T: Into<Self>> { // OK
    f1: T,
}
```

I have pushed an update, does that seem correct?